### DISCUSSION TITLE
XFRM interface address

### DISCUSSION THREAD

**Discussion by edo1** [strongswan/strongswan#1582] [Q&A]
Here is my swanctl.conf:
```
connections {
    test {
        remote_addrs = 1.2.3.4

        remote {
            auth = pubkey
            id = test.com
        }
        local {
            auth = eap
            eap_id = test
        }
        version = 2
        vips = 0.0.0.0,::
        children {
            home {
                local_ts = 0.0.0.0/0,::/0
                remote_ts = 0.0.0.0/0,::/0
                start_action = trap
                mode = tunnel
                if_id_in=123
                if_id_out=123
            }
        }
    }
}
```

Here is my XFRM interface:
```
# ip -d link show dev ipsec0
7: ipsec0@lo: <NOARP,UP,LOWER_UP> mtu 1400 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/none  promiscuity 0 minmtu 68 maxmtu 65535 
    xfrm if_id 0x7b addrgenmode random numtxqueues 1 numrxqueues 1 gso_max_size 65536 gso_max_segs 65535 
```

From the remote host, I received the address 100.100.247.5, but it is assigned to another interface:
```
# ip addr|grep 100.100
    inet 100.100.247.5/32 scope global ppp4
```
and the tunnel doesn't work.

If I do
```
# ip addr del 100.100.247.5/32 dev ppp4
# ip addr add 100.100.247.5/16 dev ipsec0
```
then the tunnel works as expected.

Do I understand correctly that the address 100.100.247.5 should be assigned to the XFRM interface? Why is it assigned to another interface?

**Comment by tobiasbrunner** [comment]
> Do I understand correctly that the address 100.100.247.5 should be assigned to the XFRM interface?

That depends on the routes you install via XFRM interface. If you install them so they reference the virtual IP (`src 100.100.247.5`), it doesn't matter on which interface that IP is located (with 5.9.10 you could even let strongSwan install such routes for you automatically, otherwise install them in an updown or vici script if you don't know the virtual IP beforehand).

If the routes don't reference the IP, i.e. they just list a destination and the XFRM interface, the address has to be installed on that interface so the correct source IP is selected. If the source IP doesn't match the IPsec policies, the packet won't be tunneled even if they are routed via XFRM interface.

> Why is it assigned to another interface?

The default is to install virtual IPs on the interface over which the peer is reachable (which is not the XFRM interface). To control this, configure `charon.install_virtual_ip_on` (you can set it to the XFRM interface as long as that exists before the daemon installs the virtual IPs).

**Reply by celesteking** [reply]
> That depends on the routes you install via XFRM interface. 

I like how the responsibility got shifted to this poor user. He didn't install anything, it was the software that installed the routes for him. He doesn't even have an updown script there. He's clueless about what the software is doing.

> it doesn't matter on which interface that IP is located 

Except for 1 small detail, which is the interface ID, `if_id` in `ip xfrm pol`, and having the routes on wrong interface (which is the default as I see) leads to traffic not being trapped by ipsec policy and getting out (or trying to) in plaintext out of the default interface. 

The whole thing needs to be reconsidered. Either install VIP on if_id (but then, how do you deal with `in != out` case?) or ask the user to configure an updown script with proper routes.

Unless I misunderstood something, of course.

**Reply by celesteking** [reply]
I just realized that an IP doesn't have to belong to an interface for a route to be able to reference said interface and said IP. Didn't know that was possible in linux.

In other words, the XFRM interface doesn't even need any IP assigned, you just need to create the interface and bring it up, the rest will be done by swan if you specify `charon.plugins.kernel-netlink.install_routes_xfrmi=yes`.

**Comment by edo1** [comment]
> To control this, configure `charon.install_virtual_ip_on`

It works for me, thanks. Could this be done using swanctl.conf?

**Reply by tobiasbrunner** [reply]
No, currently not.
